File: android/src/main/java/com/reactnativereadium/utils/NitroTypeConverters.kt
package com.reactnativereadium.utils

import android.graphics.Color
import com.margelo.nitro.reactnativereadium.*
import org.readium.r2.navigator.epub.EpubPreferences as ReadiumEpubPreferences
import org.readium.r2.navigator.preferences.ColumnCount
import org.readium.r2.navigator.preferences.Fit
import org.readium.r2.navigator.preferences.FontFamily
import org.readium.r2.navigator.preferences.ImageFilter
import org.readium.r2.navigator.preferences.ReadingProgression as NavReadingProgression
import org.readium.r2.navigator.preferences.Spread as NavSpread
import org.readium.r2.navigator.preferences.TextAlign
import org.readium.r2.navigator.preferences.Theme
import org.readium.r2.navigator.preferences.Color as ReadiumColor
import org.readium.r2.navigator.preferences.Axis
import org.readium.r2.shared.util.Language
import org.readium.r2.shared.publication.services.search.SearchService
import org.readium.r2.navigator.DecorableNavigator
import org.readium.r2.navigator.Navigator
import org.readium.r2.navigator.SelectableNavigator
import org.readium.r2.navigator.epub.EpubNavigatorFactory
import org.readium.r2.navigator.epub.EpubNavigatorFragment
import org.readium.r2.navigator.pdf.PdfNavigatorFactory
import org.readium.r2.navigator.pdf.PdfNavigatorFragment
import org.readium.r2.shared.publication.Publication
import org.readium.r2.shared.publication.services.search.SearchService as SearchServiceClass

import org.readium.r2.shared.publication.Locator as ReadiumLocator
import org.readium.r2.shared.publication.Link as ReadiumLink
import org.readium.r2.shared.publication.Metadata as ReadiumMetadata
import org.readium.r2.shared.util.Url as ReadiumUrl
import org.readium.r2.shared.util.mediatype.MediaType as ReadiumMediaType
import org.readium.r2.navigator.Decoration as ReadiumDecoration

import org.readium.adapter.pdfium.navigator.PdfiumPreferences as ReadiumPdfPreferences
import org.readium.adapter.pdfium.navigator.PdfiumEngineProvider

import android.util.Log

private const val TAG = "ReadiumConverters"


// MARK: - Nitro → Readium converters

// Sepia theme colors — unified across iOS, Android, and web
private const val SEPIA_BACKGROUND = "#f4ecd8"
private const val SEPIA_TEXT = "#5f4b32"

internal fun nitroPreferencesToEpub(prefs: Preferences): ReadiumEpubPreferences {
  // When theme is sepia and no explicit colors are set, inject our unified
  // sepia colors so the result matches iOS and web exactly.
  val bgColor = prefs.backgroundColor?.let { parseReadiumColor(it) }
    ?: if (prefs.theme == "sepia" && prefs.backgroundColor == null) parseReadiumColor(SEPIA_BACKGROUND) else null
  val txtColor = prefs.textColor?.let { parseReadiumColor(it) }
    ?: if (prefs.theme == "sepia" && prefs.textColor == null) parseReadiumColor(SEPIA_TEXT) else null

  return try {
    ReadiumEpubPreferences(
      backgroundColor = bgColor,
      columnCount = prefs.columnCount?.let { parseColumnCount(it) },
      fontFamily = prefs.fontFamily?.let { FontFamily(it) },
      fontSize = prefs.fontSize,
      fontWeight = prefs.fontWeight,
      hyphens = prefs.hyphens,
      imageFilter = prefs.imageFilter?.let { parseImageFilter(it) },
      language = prefs.language?.let { Language(it) },
      letterSpacing = prefs.letterSpacing,
      ligatures = prefs.ligatures,
      lineHeight = prefs.lineHeight,
      pageMargins = prefs.pageMargins,
      paragraphIndent = prefs.paragraphIndent,
      paragraphSpacing = prefs.paragraphSpacing,
      publisherStyles = prefs.publisherStyles,
      readingProgression = prefs.readingProgression?.let { parseReadingProgression(it) },
      scroll = prefs.scroll,
      spread = prefs.spread?.let { parseSpread(it) },
      textAlign = prefs.textAlign?.let { parseTextAlign(it) },
      textColor = txtColor,
      textNormalization = prefs.textNormalization,
      theme = prefs.theme?.let { parseTheme(it) },
      typeScale = prefs.typeScale,
      verticalText = prefs.verticalText,
      wordSpacing = prefs.wordSpacing,
    )
  } catch (e: IllegalArgumentException) {
    Log.e(TAG, "Failed to build EPUB preferences, using defaults", e)
    ReadiumEpubPreferences()
  }
}

internal fun nitroPreferencesToPdf(prefs: Preferences): ReadiumPdfPreferences {
  return try {
    ReadiumPdfPreferences(
      fit = prefs.fit?.let { parseFit(it) },
      pageSpacing = prefs.pageSpacing,
      readingProgression = prefs.readingProgression?.let { parseReadingProgression(it) },
      scrollAxis = prefs.scrollAxis?.let { parseAxis(it) },
    )
  } catch (e: IllegalArgumentException) {
    Log.e(TAG, "Failed to build PDF preferences, using defaults", e)
    ReadiumPdfPreferences()
  }
}

private fun parseReadiumColor(hex: String): ReadiumColor? {
  return try {
    ReadiumColor(android.graphics.Color.parseColor(hex))
  } catch (e: Exception) {
    null
  }
}

private fun parseTheme(value: String): Theme? = when (value.lowercase()) {
  "light" -> Theme.LIGHT
  "dark" -> Theme.DARK
  "sepia" -> Theme.SEPIA
  else -> null
}

private fun parseColumnCount(value: String): ColumnCount? = when (value.lowercase()) {
  "auto" -> ColumnCount.AUTO
  "1" -> ColumnCount.ONE
  "2" -> ColumnCount.TWO
  else -> null
}

private fun parseImageFilter(value: String): ImageFilter? = when (value.lowercase()) {
  "darken" -> ImageFilter.DARKEN
  "invert" -> ImageFilter.INVERT
  else -> null
}

private fun parseReadingProgression(value: String): NavReadingProgression? = when (value.lowercase()) {
  "ltr" -> NavReadingProgression.LTR
  "rtl" -> NavReadingProgression.RTL
  else -> null
}

private fun parseSpread(value: String): NavSpread? = when (value.lowercase()) {
  "auto" -> NavSpread.AUTO
  "never" -> NavSpread.NEVER
  "always" -> NavSpread.ALWAYS
  else -> null
}

private fun parseTextAlign(value: String): TextAlign? = when (value.lowercase()) {
  "center" -> TextAlign.CENTER
  "justify" -> TextAlign.JUSTIFY
  "start" -> TextAlign.START
  "end" -> TextAlign.END
  "left" -> TextAlign.LEFT
  "right" -> TextAlign.RIGHT
  else -> null
}

private fun parseFit(value: String): Fit? = when (value.lowercase()) {
  "cover" -> Fit.COVER
  "contain" -> Fit.CONTAIN
  "width" -> Fit.WIDTH
  "height" -> Fit.HEIGHT
  else -> null
}

/**
 * Stands in for an href that named no resource, so the locator survives conversion
 * long enough for `ReaderService` to remap it against the opened publication.
 */
internal const val UNRESOLVED_HREF_PLACEHOLDER = "__readium_unresolved__"

internal fun nitroLocatorToReadium(loc: Locator): ReadiumLocator? {
  val normalized = normalizeHref(loc.href)
  // Older web builds persisted PDF locators with a fragment-only href such as
  // "#page=3". Keep them rather than dropping the reading position on the floor.
  val resourcePath = normalized.resourcePath.ifEmpty { UNRESOLVED_HREF_PLACEHOLDER }
  val href = ReadiumUrl(resourcePath) ?: return null
  val mediaType = ReadiumMediaType(loc.type) ?: ReadiumMediaType.BINARY

  // Merge any fragment from the href into locations.fragments
  val fragments = buildList {
    normalized.fragment?.let { add(it) }
  }

  return ReadiumLocator(
    href = href,
    mediaType = mediaType,
    title = loc.title,
    locations = ReadiumLocator.Locations(
      fragments = fragments,
      progression = loc.locations?.progression,
      position = loc.locations?.position?.toInt(),
      totalProgression = loc.locations?.totalProgression
    ),
    text = ReadiumLocator.Text(
      before = loc.text?.before,
      highlight = loc.text?.highlight,
      after = loc.text?.after
    )
  )
}

internal fun nitroDecorationToReadium(dec: Decoration): ReadiumDecoration? {
  val locator = nitroLocatorToReadium(dec.locator) ?: return null

  val style: ReadiumDecoration.Style = when (dec.style.type) {
    "highlight" -> ReadiumDecoration.Style.Highlight(
      tint = parseColorString(dec.style.tint),
      isActive = dec.style.isActive ?: false
    )
    "underline" -> ReadiumDecoration.Style.Underline(
      tint = parseColorString(dec.style.tint),
      isActive = dec.style.isActive ?: false
    )
    else -> return null
  }

  val extras: Map<String, Any> = dec.extras ?: emptyMap()

  return ReadiumDecoration(
    id = dec.id,
    locator = locator,
    style = style,
    extras = extras
  )
}

private fun parseAxis(value: String): Axis? = when (value) {
  "horizontal" -> Axis.HORIZONTAL
  "vertical" -> Axis.VERTICAL
  else -> null
}

internal fun parseColorString(colorString: String?): Int {
  if (colorString == null) return Color.YELLOW
  val trimmed = colorString.trim()
  return try {
    when {
      trimmed.startsWith("rgb(") -> {
        val values = trimmed.substringAfter("(").substringBefore(")").split(",")
        Color.rgb(values[0].trim().toInt(), values[1].trim().toInt(), values[2].trim().toInt())
      }
      trimmed.startsWith("rgba(") -> {
        val values = trimmed.substringAfter("(").substringBefore(")").split(",")
        Color.argb((values[3].trim().toFloat() * 255).toInt(),
          values[0].trim().toInt(), values[1].trim().toInt(), values[2].trim().toInt())
      }
      else -> Color.parseColor(trimmed) // handles hex + named colors
    }
  } catch (e: Exception) {
    Color.YELLOW
  }
}

// MARK: - Readium → Nitro converters

internal fun readiumLocatorToNitro(loc: ReadiumLocator): Locator {
  val locations = LocatorLocations(
    progression = loc.locations.progression,
    position = loc.locations.position?.toDouble(),
    totalProgression = loc.locations.totalProgression
  )

  val text = if (loc.text.highlight != null || loc.text.before != null || loc.text.after != null) {
    LocatorText(
      before = loc.text.before,
      highlight = loc.text.highlight,
      after = loc.text.after
    )
  } else null

  return Locator(
    href = loc.href.toString(),
    type = loc.mediaType.toString(),
    target = null,
    title = loc.title,
    locations = locations,
    text = text
  )
}

internal fun readiumLinkToNitro(link: ReadiumLink, depth: Double = 0.0, parentHref: String? = null, position: Double = 0.0): Link {
  return Link(
    href = link.href.toString(),
    title = link.title,
    rels = link.rels.toTypedArray(),
    languages = link.languages.toTypedArray(),
    depth = depth,
    hasChildren = if (link.children.isNotEmpty()) true else null,
    parentHref = parentHref,
    position = position
  )
}

internal fun flattenReadiumLinks(links: List<ReadiumLink>, depth: Double = 0.0, parentHref: String? = null): List<Link> {
  val result = mutableListOf<Link>()
  for ((index, link) in links.withIndex()) {
    result.add(readiumLinkToNitro(link, depth, parentHref, index.toDouble()))
    if (link.children.isNotEmpty()) {
      result.addAll(flattenReadiumLinks(link.children, depth + 1, link.href.toString()))
    }
  }
  return result
}

internal fun readiumDecorationToNitro(dec: ReadiumDecoration): Decoration {
  val locator = readiumLocatorToNitro(dec.locator)

  val style = DecorationStyle(
    type = when (dec.style) {
      is ReadiumDecoration.Style.Highlight -> "highlight"
      is ReadiumDecoration.Style.Underline -> "underline"
      else -> "custom"
    },
    tint = when (dec.style) {
      is ReadiumDecoration.Style.Highlight -> colorToHex((dec.style as ReadiumDecoration.Style.Highlight).tint)
      is ReadiumDecoration.Style.Underline -> colorToHex((dec.style as ReadiumDecoration.Style.Underline).tint)
      else -> null
    },
    isActive = when (dec.style) {
      is ReadiumDecoration.Style.Highlight -> (dec.style as ReadiumDecoration.Style.Highlight).isActive
      is ReadiumDecoration.Style.Underline -> (dec.style as ReadiumDecoration.Style.Underline).isActive
      else -> null
    },
    id = null, html = null, css = null, layout = null, width = null
  )

  val extras = if (dec.extras.isNotEmpty()) {
    dec.extras.entries.associate { it.key to it.value.toString() }
  } else null

  return Decoration(id = dec.id, locator = locator, style = style, extras = extras)
}

internal fun readiumMetadataToNitro(meta: ReadiumMetadata): PublicationMetadata {
  fun contributors(list: List<org.readium.r2.shared.publication.Contributor>): Array<Contributor>? {
    if (list.isEmpty()) return null
    return list.map {
      Contributor(
        name = it.name,
        sortAs = null,
        identifier = it.identifier,
        role = it.roles.firstOrNull(),
        position = it.position
      )
    }.toTypedArray()
  }

  fun subjects(list: List<org.readium.r2.shared.publication.Subject>): Array<Subject>? {
    if (list.isEmpty()) return null
    return list.map {
      Subject(name = it.name, sortAs = null, code = it.code, scheme = it.scheme)
    }.toTypedArray()
  }

  return PublicationMetadata(
    title = meta.title ?: "Untitled",
    sortAs = null,
    subtitle = meta.localizedSubtitle?.string,
    identifier = meta.identifier,
    conformsTo = meta.conformsTo.map { it.uri }.toTypedArray(),
    accessibility = null,
    modified = meta.modified?.toString(),
    published = meta.published?.toString(),
    language = if (meta.languages.isNotEmpty()) meta.languages.toTypedArray() else null,
    author = contributors(meta.authors),
    translator = contributors(meta.translators),
    editor = contributors(meta.editors),
    artist = contributors(meta.artists),
    illustrator = contributors(meta.illustrators),
    letterer = contributors(meta.letterers),
    penciler = contributors(meta.pencilers),
    colorist = contributors(meta.colorists),
    inker = contributors(meta.inkers),
    narrator = contributors(meta.narrators),
    contributor = contributors(meta.contributors),
    publisher = contributors(meta.publishers),
    imprint = contributors(meta.imprints),
    subject = subjects(meta.subjects),
    layout = meta.layout?.value,
    readingProgression = meta.readingProgression?.name?.lowercase(),
    description = meta.description,
    duration = meta.duration,
    numberOfPages = meta.numberOfPages?.toDouble(),
    belongsTo = null
  )
}

/**
 * What this reader can do with the publication it opened.
 *
 * Asks Readium rather than consulting a list: `isEffective` is the toolkit's own
 * answer for whether submitting a preference would change anything, given this
 * publication's layout and the preferences in force.
 */
internal fun readiumCapabilities(
  publication: Publication,
  navigator: Navigator,
  preferences: Preferences?
): Capabilities {

  val epubEditor = (navigator as? EpubNavigatorFragment)?.let { nav ->
    val current = preferences?.let { nitroPreferencesToEpub(it) } ?: ReadiumEpubPreferences()
    EpubNavigatorFactory(publication).createPreferencesEditor(current)
  }

  val pdfEditor = (navigator as? PdfNavigatorFragment<*, *>)?.let { nav ->
    val current = preferences?.let { nitroPreferencesToPdf(it) } ?: ReadiumPdfPreferences()
    PdfNavigatorFactory(publication, PdfiumEngineProvider()).createPreferencesEditor(current)
  }

  return Capabilities(
    // Shared Preferences
    readingProgression = epubEditor?.readingProgression?.isEffective ?: pdfEditor?.readingProgression?.isEffective ?: false,

    // EPUB-only Preferences
    backgroundColor = epubEditor?.backgroundColor?.isEffective ?: false,
    columnCount = epubEditor?.columnCount?.isEffective ?: false,
    fontFamily = epubEditor?.fontFamily?.isEffective ?: false,
    fontSize = epubEditor?.fontSize?.isEffective ?: false,
    fontWeight = epubEditor?.fontWeight?.isEffective ?: false,
    hyphens = epubEditor?.hyphens?.isEffective ?: false,
    imageFilter = epubEditor?.imageFilter?.isEffective ?: false,
    language = epubEditor?.language?.isEffective ?: false,
    letterSpacing = epubEditor?.letterSpacing?.isEffective ?: false,
    ligatures = epubEditor?.ligatures?.isEffective ?: false,
    lineHeight = epubEditor?.lineHeight?.isEffective ?: false,
    pageMargins = epubEditor?.pageMargins?.isEffective ?: false,
    paragraphIndent = epubEditor?.paragraphIndent?.isEffective ?: false,
    paragraphSpacing = epubEditor?.paragraphSpacing?.isEffective ?: false,
    publisherStyles = epubEditor?.publisherStyles?.isEffective ?: false,
    scroll = epubEditor?.scroll?.isEffective ?: false,
    spread = epubEditor?.spread?.isEffective ?: false,
    textAlign = epubEditor?.textAlign?.isEffective ?: false,
    textColor = epubEditor?.textColor?.isEffective ?: false,
    textNormalization = epubEditor?.textNormalization?.isEffective ?: false,
    theme = epubEditor?.theme?.isEffective ?: false,
    typeScale = epubEditor?.typeScale?.isEffective ?: false,
    verticalText = epubEditor?.verticalText?.isEffective ?: false,
    wordSpacing = epubEditor?.wordSpacing?.isEffective ?: false,

    // PDF-only Preferences
    fit = pdfEditor?.fit?.isEffective ?: false,
    offsetFirstPage = false,
    pageSpacing = pdfEditor?.pageSpacing?.isEffective ?: false,
    scrollAxis = pdfEditor?.scrollAxis?.isEffective ?: false,
    visibleScrollbar = false,

    // Web-only or Derived Capabilities
    zoom = false,
    search = publication.findService(SearchServiceClass::class) != null,
    decorations = navigator is DecorableNavigator,
    selection = navigator is SelectableNavigator
  )
}

internal fun colorToHex(color: Int): String = String.format("#%08X", color)

internal fun nitroSearchOptionsToReadium(options: SearchOptions): SearchService.Options {
  var result = SearchService.Options()
  options.caseSensitive?.let { result = result.copy(caseSensitive = it) }
  options.diacriticSensitive?.let { result = result.copy(diacriticSensitive = it) }
  options.wholeWord?.let { result = result.copy(wholeWord = it) }
  options.regularExpression?.let { result = result.copy(regularExpression = it) }
  options.language?.let { result = result.copy(language = it) }
  return result
}

internal fun nitroSearchResultFromReadium(locator: ReadiumLocator): SearchResult =
  SearchResult(
    locator = readiumLocatorToNitro(locator),
    before = locator.text.before,
    highlight = locator.text.highlight,
    after = locator.text.after
  )

File: ios/NitroTypeConverters.swift
import Foundation
import NitroModules
import ReadiumShared
import ReadiumNavigator
import UIKit

// Nitro-generated types shadow Readium SDK types of the same name.
// Use these aliases when referring to the Readium SDK types explicitly.
typealias RDecoration = ReadiumNavigator.Decoration
typealias RLocator = ReadiumShared.Locator
typealias RLink = ReadiumShared.Link

// MARK: - Nitro → Readium converters

func nitroPreferencesToEPUB(_ prefs: Preferences) -> EPUBPreferences {
  return EPUBPreferences(
    backgroundColor: prefs.backgroundColor.flatMap { ReadiumNavigator.Color(hex: $0) },
    columnCount: prefs.columnCount.flatMap { ColumnCount(rawValue: $0) },
    fontFamily: prefs.fontFamily.map { FontFamily(rawValue: $0) },
    fontSize: prefs.fontSize,
    fontWeight: prefs.fontWeight,
    hyphens: prefs.hyphens,
    imageFilter: prefs.imageFilter.flatMap { ImageFilter(rawValue: $0) },
    language: prefs.language.map { Language(code: .bcp47($0)) },
    letterSpacing: prefs.letterSpacing,
    ligatures: prefs.ligatures,
    lineHeight: prefs.lineHeight,
    pageMargins: prefs.pageMargins,
    paragraphIndent: prefs.paragraphIndent,
    paragraphSpacing: prefs.paragraphSpacing,
    publisherStyles: prefs.publisherStyles,
    readingProgression: prefs.readingProgression.flatMap { ReadiumNavigator.ReadingProgression(rawValue: $0) },
    scroll: prefs.scroll,
    spread: prefs.spread.flatMap { Spread(rawValue: $0) },
    textAlign: prefs.textAlign.flatMap { TextAlignment(rawValue: $0) },
    textColor: prefs.textColor.flatMap { ReadiumNavigator.Color(hex: $0) },
    textNormalization: prefs.textNormalization,
    theme: prefs.theme.flatMap { Theme(rawValue: $0) },
    typeScale: prefs.typeScale,
    verticalText: prefs.verticalText,
    wordSpacing: prefs.wordSpacing
  )
}

func nitroLocatorToReadium(_ loc: Locator) -> RLocator? {
  let locatorData = LocatorData(
    href: loc.href,
    type: loc.type,
    title: loc.title,
    locations: loc.locations.map {
      LocationsData(progression: $0.progression, position: $0.position.map { Int($0) }, totalProgression: $0.totalProgression)
    },
    text: loc.text.map {
      TextData(before: $0.before, highlight: $0.highlight, after: $0.after)
    }
  )
  return locatorData.toLocator()
}

func nitroDecorationToReadium(_ dec: Decoration) -> RDecoration? {
  guard let readiumLocator = nitroLocatorToReadium(dec.locator) else { return nil }

  let styleData = StyleData(type: dec.style.type, tint: dec.style.tint, isActive: dec.style.isActive)
  guard let readiumStyle = styleData.toDecorationStyle() else { return nil }

  var userInfo: [AnyHashable: AnyHashable] = [:]
  if let extras = dec.extras {
    for (key, value) in extras {
      userInfo[key] = value
    }
  }

  return RDecoration(
    id: dec.id,
    locator: readiumLocator,
    style: readiumStyle,
    userInfo: userInfo
  )
}

// MARK: - Readium → Nitro converters

func readiumDecorationToNitro(_ dec: RDecoration, group: String) -> Decoration {
  let locator = readiumLocatorToNitro(dec.locator)

  var styleType = "highlight"
  var tint: String?
  var isActive: Bool?

  if let highlightConfig = dec.style.config as? RDecoration.Style.HighlightConfig {
    styleType = "highlight"
    tint = highlightConfig.tint.map { $0.cssHex }
    isActive = highlightConfig.isActive
  }

  let style = DecorationStyle(type: styleType, tint: tint, isActive: isActive, id: nil, html: nil, css: nil, layout: nil, width: nil)

  var extras: [String: String]?
  if !dec.userInfo.isEmpty {
    var dict: [String: String] = [:]
    for (key, value) in dec.userInfo {
      if let k = key as? String {
        dict[k] = "\(value)"
      }
    }
    extras = dict
  }

  return Decoration(id: dec.id, locator: locator, style: style, extras: extras)
}

func readiumLocatorToNitro(_ loc: RLocator) -> Locator {
  let locations = LocatorLocations(
    progression: loc.locations.progression ?? 0,
    position: loc.locations.position.map { Double($0) },
    totalProgression: loc.locations.totalProgression
  )

  let text: LocatorText? = {
    let t = loc.text
    if t.after == nil && t.before == nil && t.highlight == nil { return nil }
    return LocatorText(before: t.before, highlight: t.highlight, after: t.after)
  }()

  return Locator(
    href: loc.href.string,
    type: loc.mediaType.string,
    target: nil,
    title: loc.title,
    locations: locations,
    text: text
  )
}

func readiumLinkToNitro(_ link: RLink) -> Link {
  return Link(
    href: link.href,
    templated: link.templated,
    type: link.mediaType?.string,
    title: link.title,
    rels: link.rels.map { "\($0)" },
    height: nil,
    width: nil,
    bitrate: nil,
    duration: nil,
    languages: link.languages
  )
}

func readiumMetadataToNitro(_ meta: ReadiumShared.Metadata) -> PublicationMetadata {
  func contributors(_ list: [ReadiumShared.Contributor]) -> [margelo.nitro.readium.Contributor]? {
    guard !list.isEmpty else { return nil }
    return list.map { Contributor(name: $0.name, sortAs: $0.sortAs, identifier: $0.identifier, role: nil, position: nil) }
  }

  func subjects(_ list: [ReadiumShared.Subject]) -> [margelo.nitro.readium.Subject]? {
    guard !list.isEmpty else { return nil }
    return list.map { Subject(name: $0.name, sortAs: $0.sortAs, code: $0.code, scheme: $0.scheme) }
  }

  return PublicationMetadata(
    title: meta.title ?? "Untitled",
    sortAs: meta.sortAs,
    subtitle: meta.subtitle,
    identifier: meta.identifier,
    accessibility: nil,
    modified: meta.modified?.description,
    published: meta.published?.description,
    language: meta.languages.isEmpty ? nil : meta.languages,
    author: contributors(meta.authors),
    translator: contributors(meta.translators),
    editor: contributors(meta.editors),
    artist: contributors(meta.artists),
    illustrator: contributors(meta.illustrators),
    letterer: contributors(meta.letterers),
    penciler: contributors(meta.pencilers),
    colorist: contributors(meta.colorists),
    inker: contributors(meta.inkers),
    narrator: contributors(meta.narrators),
    contributor: contributors(meta.contributors),
    publisher: contributors(meta.publishers),
    imprint: contributors(meta.imprints),
    subject: subjects(meta.subjects),
    layout: nil,
    readingProgression: meta.readingProgression.rawValue,
    description: meta.description,
    duration: meta.duration,
    numberOfPages: meta.numberOfPages.map { Double($0) },
    belongsTo: nil
  )
}
